File: examples/minimart-go-nuxt/backend/internal/infrastructure/controller/order_controller.go
package controller

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"minimart/internal/application/dto"
	"minimart/internal/application/usecase"
	"minimart/internal/domain/entity"
)

// OrderController serves POST /orders and GET /orders.
type OrderController struct {
	orders *usecase.OrderUsecase
}

// NewOrderController returns an OrderController.
func NewOrderController(orders *usecase.OrderUsecase) *OrderController {
	return &OrderController{orders: orders}
}

// Create redeems one product for points.
//
// The endpoint is intentionally not idempotent: two requests are two
// redemptions, which is the correct behaviour for a shop.
func (ctl *OrderController) Create(c *gin.Context) {
	userID, err := callerID(c)
	if err != nil {
		respondError(c, err)
		return
	}

	var req dto.CreateOrderRequest
	if err := decodeJSON(c, &req); err != nil {
		respondError(c, err)
		return
	}
	if req.ProductID == nil || *req.ProductID <= 0 {
		respondError(c, entity.ErrInvalidRequest)
		return
	}

	created, err := ctl.orders.Redeem(c.Request.Context(), userID, *req.ProductID)
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusCreated, created)
}

// ListAll returns every user's orders, newest first.
//
// There is no role check here. The route is wired behind middleware.RequireRole
// and that is the only place the rule lives: an authorisation restated in the
// handler is an authorisation that can drift from the one in front of it, and
// the drift is invisible until someone removes the wrong one.
func (ctl *OrderController) ListAll(c *gin.Context) {
	list, err := ctl.orders.ListAll(c.Request.Context())
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, list)
}

// List returns the caller's own orders, newest first.
func (ctl *OrderController) List(c *gin.Context) {
	userID, err := callerID(c)
	if err != nil {
		respondError(c, err)
		return
	}
	list, err := ctl.orders.ListMine(c.Request.Context(), userID)
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, list)
}

File: examples/minimart-java-next/backend/src/main/java/minimart/infrastructure/web/OrderController.java
package minimart.infrastructure.web;

import jakarta.servlet.http.HttpServletRequest;
import minimart.application.dto.CreateOrderRequest;
import minimart.application.dto.CreateOrderResponse;
import minimart.application.dto.OrderList;
import minimart.application.usecase.OrderUsecase;
import minimart.domain.DomainException;
import minimart.domain.ErrorCode;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

/** Serves POST /orders and GET /orders. */
@RestController
@RequestMapping("/api/v1")
public class OrderController {

  private final OrderUsecase orders;

  /** A controller over {@code orders}. */
  public OrderController(OrderUsecase orders) {
    this.orders = orders;
  }

  /**
   * Redeems one product for points.
   *
   * <p>The endpoint is intentionally not idempotent: two requests are two redemptions, which is the
   * correct behaviour for a shop.
   */
  @PostMapping("/orders")
  @ResponseStatus(HttpStatus.CREATED)
  public CreateOrderResponse create(HttpServletRequest http, Authentication caller) {
    long userId = Caller.id(caller);

    CreateOrderRequest request = JsonBody.parse(http, CreateOrderRequest.class);
    if (request.productId() == null || request.productId() <= 0) {
      throw new DomainException(ErrorCode.INVALID_REQUEST);
    }

    return orders.redeem(userId, request.productId());
  }

  /** Returns the caller's own orders, newest first. */
  @GetMapping("/orders")
  public OrderList list(Authentication caller) {
    return orders.listMine(Caller.id(caller));
  }
}
